src/appleseed/renderer/kernel/rendering/masterrenderer.cpp

```diff
@@ -237,6 +243,8 @@ struct MasterRenderer::Impl
             // Insert post-processing time into frame's render info.
             render_info.insert("post_processing_time", stopwatch.get_seconds());
             RENDERER_LOG_INFO("post-processing time: %s.", pretty_time(stopwatch.get_seconds()).c_str());
+
+            controller.on_rendering_success();
```

Comment:
There is quite a lot of clean-up happening in `initialize_and_render_frame`. I'm concerned that calling `on_rendering_success` and `on_rendering_abort` here will lead to side effects. 

The controller is an abstraction made to be able to render with appleseed from studio, blender, max, cli and such. It is very easy to introduce a bug into that.

Reply:
From what I see, these are the differences it makes to "bubble up" these calls:
* The render time will be added to the frame's render info *before* calling `on_rendering_*` (while we could call `on_rendering_abort` before it, I don't believe this is something that could lead to problems)
* For `on_rendering_success` *only*, `postprocess()` will be called before emitting it, and so will the post-processing time be added to render info

Calling `on_rendering_success` before post-processing the frame (as it's currently happening) seems erroneous.. and it's what actually leads to the bug. Maybe a frame's "lifetime events" should be split up into "rendering success" and something like "post process success"?

Reply:
> Calling on_rendering_success before post-processing the frame (as it's currently happening) seems erroneous.. and it's what actually leads to the bug. Maybe a frame's "lifetime events" should be split up into "rendering success" and something like "post process success"?

Yes, we definitely need more precision there. But it's a code not easy to edit as it impact all plugins and studio and the cli 🤣